com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java

```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
         String utf8Data = data.getString(PROTOCOL_ENCODING);
 
         try {
-            logger.fine("\n[[RESPONSE]]\n" + utf8Data);
+            if (message instanceof Messages.Request) {
+                logger.fine("\n[[REQUEST]]\n" + utf8Data);
+            } else if (message instanceof Messages.Event) {
+                logger.fine("\n[[EVENT]]\n" + utf8Data);
+            } else {
+                logger.fine("\n[[RESPONSE]]\n" + utf8Data);
+            }
             this.writer.write(utf8Data);
             this.writer.flush();
         } catch (IOException e) {
             logger.log(Level.SEVERE, String.format("Write data to io exception: %s", e.toString()), e);
         }
     }
 
+    @Override
+    public void sendEvent(DebugEvent event) {
+        sendMessage(new Messages.Event(event.type, event));
+    }
+
+    @Override
+    public void sendResponse(Messages.Response response) {
+        sendMessage(response);
+    }
+
+    @Override
+    public void sendRequest(Messages.Request request, Consumer<Messages.Response> cb) {
+        sendRequest(request, 0, cb);
+    }
+
+    @Override
+    public void sendRequest(Messages.Request request, long timeout, Consumer<Messages.Response> cb) {
+        Timer timer = new Timer();
+        Disposable[] disposable = new Disposable[1];
+        boolean[] executed = new boolean[]{false};
+        disposable[0] = subject.filter(response -> response.request_seq == request.seq).subscribe((response) -> {
```

Comment:
If we only need to get one response from the subject, we can use take(1) and we don't need to dispose the subscription that way.
subject.filter(...).take(1).subscribe(...)